src/core/server/rendering/rendering_service.tsx

```diff
@@ -0,0 +1,126 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import React from 'react';
+import { renderToStaticMarkup } from 'react-dom/server';
+import { format } from 'prettier';
+import { take } from 'rxjs/operators';
+
+import { i18n } from '@kbn/i18n';
+
+import { CoreService } from '../../types';
+import { CoreContext } from '../core_context';
+import { ensureRawRequest } from '../http/router';
+import { Template } from './views';
+import { RenderingSetupDeps, RenderingServiceSetup } from './types';
+
+/** @internal */
+export class RenderingService implements CoreService<RenderingServiceSetup> {
+  constructor(private readonly coreContext: CoreContext) {}
+
+  public async setup({
+    http,
+    legacy,
+    legacyPlugins,
+    plugins,
+  }: RenderingSetupDeps): Promise<RenderingServiceSetup> {
+    const uiPlugins = await Promise.all(
+      [...plugins.uiPlugins.public].map(async ([pluginId, plugin]) => {
+        const config = ((await plugins.uiPlugins.browserConfigs
+          .get(pluginId)
+          ?.pipe(take(1))
+          .toPromise()) ?? {}) as Record<string, any>;
+
+        return { id: pluginId, plugin, config };
+      })
+    );
+
+    return {
+      getRenderingProvider: ({ request, uiSettings, injectedVarsOverrides }) => {
+        const basePath = http.basePath.get(request);
+
+        return {
+          render: async (
+            id = 'core',
+            options: { includeUserSettings: boolean } = { includeUserSettings: true }
+          ) => {
+            const { env } = this.coreContext;
+            const settings = {
+              defaults: await uiSettings.getAll(),
+              user: options.includeUserSettings ? await uiSettings.getUserProvided() : {},
+            };
+            const vars = await legacy.getVars(
+              id,
+              ensureRawRequest(request),
+              uiPlugins.map(({ config }) => config),
+              injectedVarsOverrides
+            );
+            const metadata = {
+              strictCsp: http.csp.strict,
+              uiPublicUrl: `${basePath}/ui`,
+              bootstrapScriptUrl: `${basePath}/bundles/app/${id}/bootstrap.js`,
+              i18n: i18n.translate,
+              locale: i18n.getLocale(),
+              darkMode: settings.user?.['theme:darkMode']?.userValue
+                ? Boolean(settings.user['theme:darkMode'].userValue)
+                : false,
+              injectedMetadata: {
+                version: env.packageInfo.version,
+                buildNumber: env.packageInfo.buildNum,
+                branch: env.packageInfo.branch,
+                basePath,
+                env,
+                legacyMode: id !== 'core',
+                i18n: {
+                  translationsUrl: `${basePath}/translations/${i18n.getLocale()}.json`,
```

Comment:
Needs test

Reply:
To clarify: we should test that the computed `translationsUrl` 1) respects the basePath and 2) respects the return value of `i18n.getLocale()`